### ISSUE TITLE
kube-proxy should report events when the proxier failovers

### ISSUE THREAD

**Issue by xh4n3** [kubernetes/kubernetes#111709]
### What would you like to be added?

kube-proxy should report k8s events of critical cases instead of just logging, e.g., kube-proxy failovers to iptables proxier when it should be running ipvs proxier, see https://github.com/kubernetes/kubernetes/blob/v1.24.3/cmd/kube-proxy/app/server_others.go#L588

There might be other critical cases need sysadmin's attention, maybe we should check everywhere.

### Why is this needed?

The background is: we recently had kube-proxy’s ipvs scheduling algorithm changed, but one of the nodes didn’t have the corresponding kernel module. When the kube-proxy agent restarts, it failed to enable ipvs proxier, and failovered to iptables proxier. As a result, the ipvs rules and iptables rules are mixed together, causing network issues. I think this kind of failover is rather important, should be reported as k8s events.

**Comment by aojea** [comment]
/priority important-soon

we only use events on a few places, maybe we have to revisit our strategy about this.

``` 
grep -r Eventf pkg/proxy/
pkg/proxy/healthcheck/service_health.go:                                hcs.recorder.Eventf(
pkg/proxy/healthcheck/proxier_health.go:                        hs.recorder.Eventf(hs.nodeRef, nil, api.EventTypeWarning, "FailedToStartProxierHealthcheck", "StartKubeProxy", msg)
pkg/proxy/util/utils.go:                recorder.Eventf(
[aojea@juanan kubernetes]$ grep -r Eventf cmd/kube-proxy/
cmd/kube-proxy/app/server.go:                           s.Recorder.Eventf(s.NodeRef, nil, api.EventTypeWarning, err.Error(), "StartKubeProxy", message)
cmd/kube-proxy/app/server.go:   s.Recorder.Eventf(s.NodeRef, nil, api.EventTypeNormal, "Starting", "StartKubeProxy", "")
```

Main risk with events is when apps abuse them and start to DDoS the apiserver, so I think that we should choose carefully which events we should generate, kube-proxy failover looks a good candidate to me

More thoughts @thockin @danwinship

**Comment by arrowfeng** [comment]
I think it's a good proposal, and failover will only generate this event on startup.

**Comment by danwinship** [comment]
I'd say we shoud get rid of proxy mode failover. IIRC, when we first added the iptables proxy, we wanted it to be default, but there were some systems that couldn't run it (old kernel or iptables binary) so we made it so iptables was default but it could fall back to userspace. I'm not sure there was really ever any good reason for falling back from ipvs to iptables. It seems more likely that the idea of fallback just got copied over.

iptables-to-userspace fallback can no longer actually happen; there is code that purports to do fallback in one case, but it would only happen if your kernel was so old that it didn't support kubernetes anyway.

So we should just remove all the fallback; if you specify `--proxy-mode ipvs` and the system can't actually do IPVS, then kube-proxy should fail.
